web/src/app/admin/admin-outgoing-logs/hooks.ts

```diff
@@ -0,0 +1,70 @@
+import Fuse from 'fuse.js'
+import { useEffect, useRef, useState } from 'react'
+
+interface FuseParams<T> {
+  data: T[]
+  keys?: Fuse.FuseOptionKey[]
+  options?: Fuse.IFuseOptions<T> & CustomOptions
+}
+
+interface CustomOptions {
+  showResultsWhenNoSearchTerm?: boolean
+}
+
+interface FuseResults<T> {
+  results: Fuse.FuseResult<T>[]
+  search: string
+  setSearch: (search: string) => void
+}
+
+const defaultOptions = {
+  shouldSort: true,
+  threshold: 0.1,
+  location: 0,
+  distance: 100,
+  maxPatternLength: 32,
+  minMatchCharLength: 1,
+}
+
+const DEFAULT_QUERY = ''
+
+export function useFuse<T>({
+  data,
+  keys,
+  options = {},
+}: FuseParams<T>): FuseResults<T> {
+  const { showResultsWhenNoSearchTerm, ...fuseOptions } = options
+  const [fuseResults, setFuseResults] = useState<Fuse.FuseResult<T>[]>([])
+  const [search, setSearch] = useState(DEFAULT_QUERY)
+  const fuse = useRef<Fuse<T>>()
+
+  useEffect(() => {
+    if (!data) return
+
+    fuse.current = new Fuse(data, {
+      ...defaultOptions,
+      ...fuseOptions,
+      keys,
+    })
+  }, [search, fuse, data, keys, options])
+
+  useEffect(() => {
+    async function set(): Promise<void> {
+      if (fuse.current) {
+        setFuseResults(await fuse.current.search(search))
+      }
+    }
+    set()
+  }, [search, fuse, data])
```

Comment:
async/await doesn't seem necessary here, we're not returning the result of `set()` or using it recursively. We also need to handle cancellations or we'll get problems with unmount at the wrong time or race conditions with results being called out-of-order.

```suggestion
  useEffect(() => {
    if (!fuse.current) return

    let callback = setFuseResults
    fuse.current.search(search).then(results => callback(results))
    
    // reset callback before unmount/update
    return () => {callback = ()=>{}}
  }, [search, fuse, data])
```